### ISSUE TITLE
Breaking change on version >=1.10.1

### ISSUE THREAD

**Issue by rooftopsparrow** [Dexus/pem#135]
Hello pem!

I recently upgraded pem to 1.11.x and tests started failing because the readCertificateInfo method stopped returning the serial property, when it did before. 

I think I've isolated the change set:

```diff
-  if ((serial = certData.match(/Serial Number:\s*([^\n]*)\n/)) && serial.length > 1) {
+  if ((serial = certData.match(/\s{8}Serial Number:\n\s{12}([^\n]*)\n\s{4}\b/)) && serial.length > 1) {
          certValues.serial = serial[1];
      }
```
via: https://github.com/Dexus/pem/commit/8c2fa9da1cf41e6e39b51758060bba52c41f91ef#diff-15e2109519d58c2c9d32e9fcee20dd07R835

Is there any reason why this was changed? Am I doing something wrong to cause this not to return anymore?

Let me know if you need any more info

**Comment by Dexus** [comment]
Can you send me the cert or csr?

Von meinem iPhone gesendet

**Comment by rooftopsparrow** [comment]
I can reproduce the behavior with the tests of this project. 

Given I checkout v1.10.0 at f071135

```diff
# Jonathan at nighthawk in ~/dev/github/pem on git:f071135 ✖︎ [19:05:56]
→ git diff
diff --git a/test/pem.js b/test/pem.js
index 16ec923..f964feb 100644
--- a/test/pem.js
+++ b/test/pem.js
@@ -423,12 +423,11 @@ exports['General Tests'] = {
             pem.readCertificateInfo(certificate, function(error, data) {
                 test.ifError(error);

-                if (data.validity) {
-                    delete data.validity;
-                }
-                if (data.serial) {
-                    delete data.serial;
-                }
+                test.ok(data.validity);
+                delete data.validity;
+
+                test.ok(data.serial);
+                delete data.serial;

                 test.deepEqual(data, certInfo);
                 test.ok(fs.readdirSync('./tmp').length === 0);
```

all tests pass.

```
# Jonathan at nighthawk in ~/dev/github/pem on git:f071135 ✖︎ [19:07:02]
→ npm test

> pem@1.10.0 test /Users/Jonathan/dev/github/pem
> grunt

Running "jshint:all" (jshint) task
>> 3 files lint free.

Running "nodeunit:all" (nodeunit) task
Testing pem.js.....................................OK
>> 235 assertions passed (40732ms)

Done.
```

Meanwhile, applying the same logic to the same test on master:

```diff
# Jonathan at nighthawk in ~/dev/github/pem on git:master ✖︎ [19:11:51]
→ git diff
diff --git a/test/pem.js b/test/pem.js
index d3c8f83..8923aa3 100644
--- a/test/pem.js
+++ b/test/pem.js
@@ -424,12 +424,11 @@ exports['General Tests'] = {
       pem.readCertificateInfo(certificate, function (error, data) {
         test.ifError(error)

-        if (data.validity) {
-          delete data.validity
-        }
-        if (data.serial) {
-          delete data.serial
-        }
+        test.ok(data.validity)
+        delete data.validity
+
+        test.ok(data.serial)
+        delete data.serial

         test.deepEqual(data, certInfo)
         test.ok(fs.readdirSync('./tmp').length === 0)
```

 the tests fail:

```
# Jonathan at nighthawk in ~/dev/github/pem on git:master ✖︎ [19:11:51]
→ npm test

> pem@0.0.0-development test /Users/Jonathan/dev/github/pem
> grunt

Running "standard:app" (standard) task

Linting files...

✓ No Problems - 0 warnings

Running "nodeunit:all" (nodeunit) task
Testing pem.js................F....................
>> General Tests - Read edited cert data from certificate
>> Error: undefined == true
>>   at /Users/Jonathan/dev/github/pem/test/pem.js:430:14
>>   at fetchCertificateData (lib/pem.js:1020:3)
>>   at /Users/Jonathan/dev/github/pem/lib/pem.js:443:12
>>   at /Users/Jonathan/dev/github/pem/lib/pem.js:1214:5
>>   at done (lib/pem.js:1163:18)
>>   at ChildProcess.<anonymous> (lib/pem.js:1178:7)
>>   at emitTwo (events.js:106:13)
>>   at ChildProcess.emit (events.js:191:7)
>>   at maybeClose (internal/child_process.js:920:16)
>>   at Process.ChildProcess._handle.onexit (internal/child_process.js:230:5)

Warning: 1/241 assertions failed (20601ms) Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.
```

The stack traces' line 430 is `test.ok(data.serial)`

Let me know if I can help any more 👍

**Comment by KaiSchwarz-cnic** [comment]
```javascript
if ((serial = certData.match(/\s{8}Serial Number:\n\s{12}([^\n]*)\n\s{4}\b/)) && serial.length > 1) {
```

That's a very restrictive regular expression @Dexus. The one prior in use wasn't that restrictive. Maybe it's just a try to replace the exact count of spaces e.g. `\s{8}` or `\s{12}` etc. with `\s*` as this may differ by openssl version. Maybe that already helps here.

So probably:
```javascript
if ((serial = certData.match(/\s*Serial Number:\n\s*([^\n]*)\n\s*\b/)) && serial.length > 1) {
```
or rolling back to the old version, if no reason is known for having that changed (maybe by a PR or so).
@rooftopsparrow : could you give my regular expression a try?

**Comment by Dexus** [comment]
Yeah, maybe it's a problem that libressl and openssl are different there in the travis ci. Because the ci doesn't break yet. And local also it was ok. Can you provide the openssl version from the command line? `openssl version`

Von meinem iPhone gesendet

**Comment by rooftopsparrow** [comment]
My current version locally:

```
openssl version
OpenSSL 0.9.8zh 14 Jan 2016
```

But this also failed via the official [node v6 Docker container](https://github.com/nodejs/docker-node/blob/17c50cb300581280805a4183524fbf57840f3a7e/6.11/Dockerfile) in our CI. So I guess whatever version that is as well. I'll have to investigate that.

**Comment by rooftopsparrow** [comment]
Bad close, sorry

**Comment by Dexus** [comment]
I'll try to build a more complex CI tomorrow. And build openssl myself, in the tests, so we can sure that every version works.

**Comment by Dexus** [comment]
Hmm, crazy... on the https://travis-ci.org/Dexus/pem/builds/277666279 it looks okay, with all Openssl versions exept of 1.1.0f because there is the dh command not available

**Comment by Dexus** [comment]
Okay, looks like there are differences between openssl and libressl, so that based on which distro you use get a drop-in-replacement and the result is that some variables not can parsed.

Hope that my tests with the branch https://github.com/Dexus/pem/tree/openssl-test-version will work now.

**Comment by rooftopsparrow** [comment]
Ok cool, because I tried running the patched tests in a clean docker and it worked 🤦‍♂️ There must be some other variable contributing to this that I'm not seeing.

Let me know if I can help.

**Comment by Dexus** [comment]
I'm patching now some functions that fail between some versions of openssl/libressl.
We will see, if it then works. If you have any idear create a PR.

**Comment by Dexus** [comment]
Should be fixed with v1.11.1
